### ISSUE TITLE
add nameplate attributes, set wordlist=None for manually-entered wormhole codes

### ISSUE THREAD

**Issue by warner** [warner/magic-wormhole#26]
There's room in the nameplate list to include per-nameplate attributes (each nameplate is delivered as a dictionary, and the nameplate ID is just one key, and we can add others). The first attribute I have in mind is a "wordlist identifier", so the sender can use an alternative wordlist, and the recipient can offer tab-completion against the right list. I'm undecided as to whether this identifier should be a simple string or integer, or if it should be a hash of the wordlist (encoded in some suitable form), so it can be a bit more open-ended (and guaranteed to be the same on both sides). The attributes are not protected at all (as they necessarily get processed before the PAKE happens), so we have to be careful with them. E.g. it's probably not a good idea to include the entire wordlist in the attribute, because then an attacker can swap it out for a really short list, and a receiver who isn't paying attention might just hit "xyTABaTABRETURN" and the attacker would have a higher chance than usual of guessing the code.

For the "choose a wormhole code offline, then type it into both machines later" mode, we could improve the receiver's experience (assuming they type it in after the sender does) by setting the wordlist identifier to None, which should disable tab-completion (at least on the wordlist).

We need to think through the user experience here, since it'd be a bit weird to have the receiver's flow change depending upon whether they happened to run before, or after, the sender runs. We have more information to work with if the receiver goes second. (A related question is what the receiver should do when it sees no data in the channel it was told to use: in the machine-generated-code case, that means they probably mistyped the channel-id, and it might be good to emit a warning).

**Comment by warner** [comment]
@dkg made a good argument to me today, to *not* include the wordlist as a nameplate attribute. The problem is that an attacker can scan those attributes before committing to a particular channel, so they could selectively break connections using a given language, while leaving other connections untouched.

As annoying as the UX and engineering consequences are, it's probably better for overall security to prevent attackers from being selective about their attacks: force them to attack (or not) all connections with equal probability, rather than letting them pick and choose their victims. That means either putting the wordlist-identifier as a *mailbox* attribute (which clients can't learn until they've claimed the nameplate, at which point they're the only one who can use it, so either they get the code right or it is broken for everybody), or not using a wordlist at all.

He suggested an interesting approach to L10N. Basically you pick half a dozen *writing systems*, say (latin, cyrillic, greek, arabic, chinese, tamil), for each one you pick the most common language written with those characters, and then you find 256 words from each. The sender's code is chosen from a subset according to their language settings (`$LANG=en_US` etc). The receiver uses the whole wordlist, but once the user has typed in the first character of a word, any tab-completion is only going to show them words from the same language. So you can effectively hide all the other words from any individual recipient, without needing any in-band indicator of the wordlist being used.

The downside is that you can't localize for specific languages: (english, french, italian, spanish) are all going to get the same wordlist. But it doesn't preclude us from adding an in-band wordlist attribute later, and certainly wouldn't add any information to enable selective attacks in the meantime.

**Comment by piegamesde** [comment]
If we add the word list information to the `claimed` message (the response message when claiming a nameplate), does the selective DOS issue still apply?

**Comment by meejah** [comment]
> If we add the word list information to the `claimed` message (the response message when claiming a nameplate), does the selective DOS issue still apply?

I believe this could potentially work .. the server at least knows if >2 sides have tried to `claim` a mailbox so an attacker has to commit to that much. It's not _quite_ the same as trying a password, but I think the result is the same: they don't learn the wordlist in use without destroying the mailbox.

(The only disadvantage would be a slight delay: waiting until the `claimed` comes back before tab-completion works)

**Comment by warner** [comment]
Yeah the `claimed` response is basically where these "mailbox attributes" would be delivered.

Having a delay between sending the `claim` and tab-completion starting to work wouldn't be too bad, we already have similar race conditions with the nameplate completion: if you run `wormhole rx` and hit tab really fast, you get no completions, because we haven't yet received the `list` response, if you wait a few seconds and hit it again, you get a list to select from, if you wait a few more seconds, you get a different list (each TAB gets you completion against the current list, and also sends a request to update the list, so it can change over time as other people add/claim nameplates).

Yeah, it is a bit weird that there's a commit/claim point that you think you can backspace over but in fact cannot. I *think* I added code to the state machine to complain if the user does this, but `readline` is finicky and I don't remember if I was able to make it work. The intention is that backspacing over the post-nameplate hyphen would cause the CLI to exit with an error, or refuse to backspace, or complain somehow until you either quit or re-typed the same nameplate, or ooh even better change the completion list to include only the one committed nameplate (so tab completion on the nameplate still looks like it works).

The current python client does this exactly for this reason, to line up for completion with variable wordlists, and to hide the latency of the CLAIM/CLAIMED messages behind the user typing the wormhole code. I think it's an ok tradeoff, but "how often does this affect people?" is a good question to ask. The best datapoint I can think of is to look at the mailbox server stats (which I should find some way to publish one of these days). In the last 5 days, we've had 5770 total sessions, broken down by "mood" outcome:
* 3930 "happy" (both clients reported success)
* 1560 "pruney" (session was deleted for lack of activity)
* 113 "lonely" (never saw a second client)
* 117 "scary" (at least one client reported a PAKE mismatch)

I *think* the "I claimed the wrong channel" case would either land in "lonely" (if they bailed before submitting a code), "scary" (65535-out-of-65536ths of the times that they proceeded through to entering a code, getting it wrong), and "happy" in the 1-out-of-65536 times they got really lucky). If that's true, then this is currently happening at most 4% of the time, and probably at most 2% (only the lonely). I'd have to run some tests to be sure, though, it might land in the "pruney" bucket, which would mean it could be happening more frequently.

**Comment by meejah** [comment]
> Because as an input field, a user may reasonably expect that they may edit its value at any point in time without consequences, and the input is only committed when pressing enter.

This sounds like an interesting UX challenge -- although of course it's a bit harder when the UX is "plain text terminal".

Since it's also possible for operators to disable the `list` command I think it's worth considering the whole feature. While I do sometimes appreciate the tab-completion for the words part, I don't know that I've ever used that for the mailbox / number part. So a UX could (right now) simply never do that bit -- and indeed, _has_ to support that mode in case the operator has disabled nameplate listing entirely.

We could change our recommendations to _not_ attempt completions on the nameplate. This, then, would give a clear path for tab-completion: determine the nameplate number, then claim (locking it in) and then do tab-completion (or could be visual "correctness checking" by a GUI).

A naive terminal client might switch this to two steps: one you enter the number -- or enough of the code to have the number, i.e. the first `-` -- it moves to a new line (making it obvious you can't delete the number part). Or it could be a two-step process: `Enter number: ` followed by `Enter code:`. I can imagine a number of ways to do this in a GUI using visual clues to show the user e.g. the number can't be changed at a certain point.

Supporting copy-paste of an "entire code" should also be possible.

----

The bottom line, though is: if we let normal users decide on a different nameplate, they're effectively "trying an unbounded number of different nameplates" which is what an  attacker might want to do (e.g. to find "the one using the weird wordlist"). So I don't think we can have that feature while also preventing attackers from targeting based on wordlist choice.

I'm good with dropping completion / verification of the nameplate before `claim` is called.

**Comment by piegamesde** [comment]
n.b. to my previous post: This was already written under the assumption that the `list` command returns empty and that there is no nameplate completion. The outlined use case for corrections still holds up without that feature.

--------

Let me throw in completely different idea: what if users could query the wordlist for a nameplate without committing to it, but the server returns a pseudo-random answer for unclaimed nameplates? I think it could be made in a way such that a potential disruptor would gain no useful information from it without already knowing a valid nameplate in the first place (and, knowing *that it is valid*, so correctly guessing is not enough).

**Comment by meejah** [comment]
> what if users could query the wordlist for a nameplate without committing to it, but the server returns a pseudo-random answer for unclaimed nameplates?

Hmm, this is definitely an interesting thing to look at!

To meet the "anti-DoS" goals I guess it would have to be available only after at least `submit_permissions()` -- and `bind` since that's currently "the first thing you do".

So then I guess the proposed feature would allow any connection that's in that state to query any number (or maybe "a reasonable number") of mailboxes for their attributes. This should take care to not reveal whether the mailbox is currently allocated or not (e.g. via timing too). Once the connection does a `claim`,  no more such queries.
